### ISSUE TITLE
Panic when using the AWS authentication backend and a static role

### ISSUE THREAD

**Issue by protochron** [hashicorp/vault#23361]
**Describe the bug**
Attempting to log in with using an AWS auth backend with IAM authentication is triggering a stacktrace with a nil pointer derefrence. Ex. `vault login -method aws role=myrole`. The current Vault leader itself isn't crashing thankfully, but the auth backend is completely broken. Reverting to Vault 1.14.4 fixes the issue.

**To Reproduce**
1. Run `vault write ...` a valid AWS auth backend and a token role that references it
2. Run `vault login --method aws role=myrole`
3. See error

**Expected behavior**
`vault login --method aws role=$myrole` to work without triggering a panic on the leader.

**Environment:**
* Vault Server Version (retrieve with `vault status`): 1.15.0
* Vault CLI Version (retrieve with `vault version`): 1.15.0
* Server Operating System/Architecture: Linux aarch64

Vault server configuration file(s):

```hcl
service_registration "consul" {
  address = "[::]:8500"
}

storage "dynamodb" {
  ha_enabled = true
}

listener "tcp" {
  address     = "[::]:8200"
}

telemetry {
  prometheus_retention_time = "30s"
  disable_hostname          = true
}
```

**Additional context**
Stack trace:
```
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: 2023-09-27T23:25:13.807Z [INFO]  http: panic serving 127.0.0.1:50390: runtime error: invalid memory address or nil pointer dereference
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: goroutine 228667 [running]:
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: net/http.(*conn).serve.func1()
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /opt/hostedtoolcache/go/1.21.1/x64/src/net/http/server.go:1868 +0xb0
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: panic({0x78c0360?, 0x1020b0c0?})
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /opt/hostedtoolcache/go/1.21.1/x64/src/runtime/panic.go:920 +0x26c
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: github.com/hashicorp/vault/builtin/credential/aws.(*backend).pathLoginIamGetRoleNameCallerIdAndEntity(0x4000179140, {0xad5e5c0, 0x4005130930}, 0x4003b56000, 0x400277ddd8?)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /home/runner/work/vault/vault/builtin/credential/aws/path_login.go:326 +0xce4
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: github.com/hashicorp/vault/builtin/credential/aws.(*backend).pathLoginUpdateIam(0x4000179140, {0xad5e5c0, 0x4005130930}, 0x4003b56000, 0x40032d62a0?)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /home/runner/work/vault/vault/builtin/credential/aws/path_login.go:1356 +0x34
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: github.com/hashicorp/vault/builtin/credential/aws.(*backend).pathLoginUpdate(0x0?, {0xad5e5c0, 0x4005130930}, 0x7010104?, 0x1?)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /home/runner/work/vault/vault/builtin/credential/aws/path_login.go:610 +0x210
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: github.com/hashicorp/vault/sdk/framework.(*Backend).HandleRequest(0x4002e764b0, {0xad5e5c0, 0x4005130930}, 0x4003b56000)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /home/runner/work/vault/vault/sdk/framework/backend.go:310 +0x900
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: github.com/hashicorp/vault/builtin/plugin/v5.(*backend).HandleRequest(0x4003827540, {0xad5e5c0, 0x4005130930}, 0x4003b56000)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /home/runner/work/vault/vault/builtin/plugin/v5/backend.go:95 +0xcc
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: github.com/hashicorp/vault/vault.(*Router).routeCommon(0x400351cfc0, {0xad5e5c0, 0x4005130930}, 0x4003b56000, 0x0)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /home/runner/work/vault/vault/vault/router.go:784 +0x11fc
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: github.com/hashicorp/vault/vault.(*Router).Route(...)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /home/runner/work/vault/vault/vault/router.go:553
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: github.com/hashicorp/vault/vault.(*Core).doRouting(0x400338cc80?, {0xad5e5c0?, 0x4005130930?}, 0x4002b36070?)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /home/runner/work/vault/vault/vault/request_handling.go:922 +0x2c
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: github.com/hashicorp/vault/vault.(*Core).handleLoginRequest(0x4002fdf800, {0xad5e5c0, 0x4005130930}, 0x4003b56000)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /home/runner/work/vault/vault/vault/request_handling.go:1486 +0x91c
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: github.com/hashicorp/vault/vault.(*Core).handleCancelableRequest(0x4002fdf800, {0xad5e5c0, 0x40051308a0}, 0x4003b56000)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /home/runner/work/vault/vault/vault/request_handling.go:764 +0x112c
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: github.com/hashicorp/vault/vault.(*Core).switchedLockHandleRequest(0x4002fdf800, {0xad5e5c0, 0x40058b9290}, 0x4003b56000, 0x1)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /home/runner/work/vault/vault/vault/request_handling.go:573 +0x4cc
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: github.com/hashicorp/vault/vault.(*Core).HandleRequest(...)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /home/runner/work/vault/vault/vault/request_handling.go:530
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: github.com/hashicorp/vault/http.request(0x4005696824?, {0xad40090, 0x40058b91a0}, 0x4003ab6300, 0x4003b56000)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /home/runner/work/vault/vault/http/handler.go:934 +0x6c
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: github.com/hashicorp/vault/http.handler.handleLogical.handleLogicalInternal.func57({0xad40090, 0x40058b91a0}, 0x4003ab6300)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /home/runner/work/vault/vault/http/logical.go:375 +0x150
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: net/http.HandlerFunc.ServeHTTP(0x4002fdf800?, {0xad40090?, 0x40058b91a0?}, 0x6634?)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /opt/hostedtoolcache/go/1.21.1/x64/src/net/http/server.go:2136 +0x38
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: github.com/hashicorp/vault/http.handler.handleRequestForwarding.func36({0xad40090, 0x40058b91a0}, 0x4003ab6300)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /home/runner/work/vault/vault/http/handler.go:868 +0x240
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: net/http.HandlerFunc.ServeHTTP(0x40027774c8?, {0xad40090?, 0x40058b91a0?}, 0x2ddb4f8?)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /opt/hostedtoolcache/go/1.21.1/x64/src/net/http/server.go:2136 +0x38
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: net/http.(*ServeMux).ServeHTTP(0x40042ce0f0?, {0xad40090, 0x40058b91a0}, 0x4003ab6300)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /opt/hostedtoolcache/go/1.21.1/x64/src/net/http/server.go:2514 +0x144
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: github.com/hashicorp/vault/http.handler.wrapHelpHandler.func48({0xad40090, 0x40058b91a0}, 0x4003ab6300)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /home/runner/work/vault/vault/http/help.go:28 +0x114
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: net/http.HandlerFunc.ServeHTTP(0x40031122a0?, {0xad40090?, 0x40058b91a0?}, 0x1721f88?)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /opt/hostedtoolcache/go/1.21.1/x64/src/net/http/server.go:2136 +0x38
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: github.com/hashicorp/vault/http.handler.wrapCORSHandler.func49({0xad40090, 0x40058b91a0}, 0x4003ab6300)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /home/runner/work/vault/vault/http/cors.go:32 +0x370
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: net/http.HandlerFunc.ServeHTTP(0x4002fdf800?, {0xad40090?, 0x40058b91a0?}, 0x4004798180?)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /opt/hostedtoolcache/go/1.21.1/x64/src/net/http/server.go:2136 +0x38
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: github.com/hashicorp/vault/http.handler.rateLimitQuotaWrapping.func50({0xad40090, 0x40058b91a0}, 0x4003ab6300)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /home/runner/work/vault/vault/http/util.go:128 +0x914
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: net/http.HandlerFunc.ServeHTTP(0x4002fdf800?, {0xad40090?, 0x40058b91a0?}, 0xc13d4c56700f873c?)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /opt/hostedtoolcache/go/1.21.1/x64/src/net/http/server.go:2136 +0x38
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: github.com/hashicorp/vault/http.wrapGenericHandler.func1({0xad3ec28?, 0x4002e3e000}, 0x4003ab6000)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /home/runner/work/vault/vault/http/handler.go:453 +0xb48
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: net/http.HandlerFunc.ServeHTTP(0x4005696824?, {0xad3ec28?, 0x4002e3e000?}, 0x2e4b80?)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /opt/hostedtoolcache/go/1.21.1/x64/src/net/http/server.go:2136 +0x38
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: github.com/hashicorp/vault/http.handler.PrintablePathCheckHandler.func51({0xad3ec28, 0x4002e3e000}, 0x4003ab6000)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /home/runner/go/pkg/mod/github.com/hashicorp/go-cleanhttp@v0.5.2/handlers.go:42 +0x88
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: net/http.HandlerFunc.ServeHTTP(0x10?, {0xad3ec28?, 0x4002e3e000?}, 0x4002e3e000?)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /opt/hostedtoolcache/go/1.21.1/x64/src/net/http/server.go:2136 +0x38
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: net/http.serverHandler.ServeHTTP({0xad05728?}, {0xad3ec28?, 0x4002e3e000?}, 0x6?)
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /opt/hostedtoolcache/go/1.21.1/x64/src/net/http/server.go:2938 +0xbc
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: net/http.(*conn).serve(0x40050b8000, {0xad5e5c0, 0x4003174ea0})
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /opt/hostedtoolcache/go/1.21.1/x64/src/net/http/server.go:2009 +0x518
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]: created by net/http.(*Server).Serve in goroutine 85
Sep 27 23:25:13 ip-10-10-2-143 vault[1319]:         /opt/hostedtoolcache/go/1.21.1/x64/src/net/http/server.go:3086 +0x4cc
```

**Comment by austingebauer** [comment]
Thanks for reporting this, @protochron. I was able to reproduce the panic but only when I delete the [client configuration](https://developer.hashicorp.com/vault/api-docs/auth/aws#configure-client) and then attempt to log in. Your reproduction step 1 seems to indicate that you wrote a client configuration, didn't delete it, and are still seeing the panic. Can you give more details as to how you're configuring the auth method?

**Comment by protochron** [comment]
@austingebauer thanks for jumping on this so quickly! In my case I don't have a client configuration because the Vault instance is using an instance profile to pull AWS credentials.

**Comment by Rumbles** [comment]
Hi, I just hit something similar to this, I can't be sure it's the same issue, but some time after updating to `1.15.0` my nomad machines were unable to connect and in the logs I see:

```
Sep 29 08:21:48 ip-10-96-2-60 systemd[1]: nomad.service: Main process exited, code=exited, status=1/FAILURE
Sep 29 08:21:53 ip-10-96-2-60 systemd[1]: nomad.service: Failed with result 'exit-code'.
Sep 29 08:21:53 ip-10-96-2-60 systemd[1]: nomad.service: Scheduled restart job, restart counter is at 84.
Sep 29 08:21:53 ip-10-96-2-60 systemd[1]: Stopped nomad.service.
Sep 29 08:21:53 ip-10-96-2-60 systemd[1]: Starting nomad.service...
Sep 29 08:21:53 ip-10-96-2-60 systemd[1]: Started nomad.service.
Sep 29 08:21:53 ip-10-96-2-60 dbus-daemon[694538]: [session uid=0 pid=694536] SELinux support is enabled
Sep 29 08:21:53 ip-10-96-2-60 bash[694529]: Error authenticating: Put "https://vault.service.consul.staging.mesos.zone:8200/v1/auth/aws/login": stream error: stream ID 1; INTERNAL_ERROR; received from peer
Sep 29 08:21:53 ip-10-96-2-60 bash[694525]: Could not login to Vault using AWS auth method
```

Killing all the vault 1.15 servers and replacing with 1.14.3 seems to have fixed the issue for me. This looks like a major show stopping bug for us

**Comment by austingebauer** [comment]
A workaround until we can get this fixed in 1.15.1 is to write an empty configuration like:

```sh
vault write -f auth/aws/config/client
```

This will be included in the known issues document for upgrading to 1.15.0.
